### DISCUSSION TITLE
Cannot mount a volume when using another WSL2 distribution

### DISCUSSION THREAD

**Discussion by alexiswl** [containers/podman#22241] [General]
### Issue Description

I have followed the instructions at https://podman-desktop.io/docs/podman/accessing-podman-from-another-wsl-instance, 

```
ls -lsrth /usr/local/bin/podman*
41M -rwxr-xr-x 1 alexiswl alexiswl 41M Feb  2 01:09 /usr/local/bin/podman-remote-static-linux_amd64
  0 lrwxrwxrwx 1 root     root      32 Apr  2 12:26 /usr/local/bin/podman -> podman-remote-static-linux_amd64
```

```
$ podman --version
podman version 4.9.1
```

```
$ podman images
REPOSITORY                TAG         IMAGE ID      CREATED      SIZE
docker.io/library/python  3-slim      23cd4350f4bd  11 days ago  135 MB
docker.io/library/ubuntu  latest      ca2b0f26964c  4 weeks ago  80.4 MB
```

```
$ podman system connection list
Name                         URI                                                                 Identity    Default
podman-machine-default-root  unix:///wsl/podman-sockets/podman-machine-default/podman-user.sock              true
```

*note the name is wrong above and because I mount my wsl2 distribution from / rather than `/mnt` `unix:///mnt/wsl` becomes `unix:///wsl`.  

I can also run and enter a container like so

```
podman run --rm -it ubuntu
root@f568beb07418:/# cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
````

However, when I run 

```
podman run --volume $PWD/inputs/:/data/inputs/ ubuntu:latest
```

I get

```
Error: statfs /home/alexiswl/cwltool-podman-test/inputs: no such file or directory
```

But this directory does exist `ls $PWD/inputs/` returns the list of files in this directory.  

How can I mount volumes from another WSL2 instance?

### Steps to reproduce the issue

Steps to reproduce the issue
1. Follow steps in `https://podman-desktop.io/docs/podman/accessing-podman-from-another-wsl-instance`
2. Check podman connection
3. Try mounting a volume like so `podman run --volume $PWD/inputs/:/data/inputs/ ubuntu:latest`

### Describe the results you received

Describe the results you received

I received the error message `Error: statfs /home/alexiswl/cwltool-podman-test/inputs: no such file or directory`

### Describe the results you expected

To run the container with the volume mounted	

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.33.5
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.10-1.fc39.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.10, commit: '
  cpuUtilization:
    idlePercent: 98.95
    systemPercent: 0.49
    userPercent: 0.56
  cpus: 8
  databaseBackend: sqlite
  distribution:
    distribution: fedora
    variant: container
    version: "39"
  eventLogger: journald
  freeLocks: 2047
  hostname: Win11Laptop2310
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
  kernel: 5.15.146.1-microsoft-standard-WSL2
  linkmode: dynamic
  logDriver: journald
  memFree: 14644989952
  memTotal: 16695500800
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.10.0-1.fc39.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.10.0
    package: netavark-1.10.3-1.fc39.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.10.3
  ociRuntime:
    name: crun
    package: crun-1.14.4-1.fc39.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.14.4
      commit: a220ca661ce078f2c37b38c92e66cf66c012d9c1
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20240220.g1e6f92b-1.fc39.x86_64
    version: |
      pasta 0^20240220.g1e6f92b-1.fc39.x86_64
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.2-1.fc39.x86_64
    version: |-
      slirp4netns version 1.2.2
      commit: 0ee2d87523e906518d34a6b423271e4826f71faf
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 4294967296
  swapTotal: 4294967296
  uptime: 1h 4m 0.00s (Approximately 0.04 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /home/user/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 0
    stopped: 1
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/user/.local/share/containers/storage
  graphRootAllocated: 1081101176832
  graphRootUsed: 1347416064
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 2
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/user/.local/share/containers/storage/volumes
version:
  APIVersion: 4.9.3
  Built: 1708357294
  BuiltTime: Tue Feb 20 02:41:34 2024
  GitCommit: ""
  GoVersion: go1.21.7
  Os: linux
  OsArch: linux/amd64
  Version: 4.9.3
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

**Comment by Luap99** [comment]
volumes are always server side, so you must make sure you volume exists on the podman machine wsl instance not on your client.

**Reply by alexiswl** [reply]
Thanks @Luap99, is there a quick example on how to mount one's WSL2 Ubuntu distro /home directory onto the /home directory of the podman server WSL2?

Alexis.

**Reply by alexiswl** [reply]
The end use-case is being able to run cwltool through WSL2 on a Windows machine

**Reply by uxduck** [reply]
Isn't this issue/discussion a blocker for anyone trying to mount files in WSL into containers? I'm surprised there isn't much presence here

**Reply by ongun-kanat-egym** [reply]
Docker desktop somehow manages to mount things from other WSL2 distros to the container. The mounts are native bind mounts. So they are somehow finding their way around mounting things accross namespaces in the Linux kernel. There must be a way to funnel the mount point from a foreign distro to podman-machine and then mount it there.

My company is currently switching to Podman but this is a bummer for my team who has hybrid WSL2 and Windows setups and would like to use a single daemon / cache of containers.

**Reply by guba-launikda** [reply]
@ongun-kanat-egym My organization is trying to do the same with a similar hybrid environment, so I am curious to see if you were able to solve this issue?

**Reply by deivard** [reply]
I have the same issues. Has anyone found a solution for mounting in Windows/WSL2 hybrid setups?

**Reply by jankap** [reply]
Wow! That's a massive bummer. That breaks my complete team's (and I bet many teams') WSL workflows. TBH I was expecting that this works out of the box, identically to Docker Desktop.

Why does `podman run --rm -it --mount type=bind,source=/home,target=/workspaces ubuntu` mount the podman machine's home, but ` podman run --rm -it --mount type=bind,source=.,target=/workspaces ubuntu` in WSL mounts the / of podman machine, and `podman run --rm -it --mount type=bind,source=.,target=/workspaces ubuntu` in Windows mounts correctly the current directory in Windows?

Oh wow shit, I need to completely revise my companie's strategy.

Any goals to change that behavior? It's absolutely essential when you want to use devcontainers in WSL, which is a very valid usecase IMO.

**Comment by rdbisme** [comment]
Somewhere else in some other discussions, people mentioned `sshfs`. Can't it be used to dynamically mount the volume? `podman-remote` communicates with the machine over ssh, doesn't it?

**Comment by rgaiacs** [comment]
My work around to this issue was to delete the WSL `podman-machine` and install Podman in the my default WSL. And to have Podman Desktop working with it, I can configure SSH.

**Reply by d3latorre** [reply]
How did u manage to make podman desktop to work with the instillation of podman in your WSL instance?, im trying to do the same, to install podman on WSL and to make it work seamlessly across both podman desktop and Linux environment

**Reply by rdbisme** [reply]
My solution: 
- `scoop install podman`
- In powershell `podman-machine init` (you want to put `podman machine start` behind a Schedule Task at boot)
- I have an Archlinux WSL
- `sudo pacman -S podman`
- In my `env` on WSL I have set ` echo ${CONTAINER_HOST} unix:///mnt/wsl/podman-sockets/podman-machine-default/podman-user.sock`
- This will make the `podman` inside WSL use the `socket` of the `podman-machine`. The `podman` on windows will also use that by default
- On Windows do `podman pull busybox:stable`
- On WSL do `podman image ls` > You should see the downloaded image. 
- The interoperability works both ways.

**Reply by d3latorre** [reply]
yes, but this does not work in case you want to mount volumes inside WSL, or does it?.
Ex: if i want to run this command in your Arch Linux WSL `podman run --rm -v ./test-data:/data:Z volume-test` it will search for the path inside the podman machine or try to resolve it in windows? or it fails, bc with my current config using this part of the documentation [here](https://podman-desktop.io/docs/podman/accessing-podman-from-another-wsl-instance) it fails.
Ive trying to find a workarround the most reliable one would be use the WSL instance that i already have as the podman machine but not sure how to configure it

**Reply by rgaiacs** [reply]
Hi @d3latorre,

> How did u manage to make podman desktop to work with the instillation of podman in your WSL instance?

Here is a more detailed step by step.

1. Create a SSH key on Windows. On PowerShell, you can use something like `ssh-keygen -t ed25519 -C "your_email@example.com"`. The SSH key will be created in `C:\User\you\.ssh`.
2. Copy the public SSH key that you created to the WSL `/home/you/.ssh/authorized_keys`.
3. Install the SSH server on your WSL. On Arch Linux, it should be something like `pacman -Syu openssh`.
4. You need to configure the SSH server to answer to your desired port. Remember that Windows will map the Windows port with the WSL port by default.
5. Test that you can access the WSL from Windows using SSH. With PowerShell, you can use `ssh you@localhost:port`.
6. Now you can configure the Podman client on Windows to talk with the Podman process in the WSL. On PowerShell, you can run something like `podman system connection add my-remote-machine --identity ~/.ssh/id_ed25519 ssh://myuser@my-server-ip/run/user/1000/podman/podman.sock`.

For more details, please reference to the official documentation at https://podman-desktop.io/docs/podman/podman-remote#prerequisites.

**Comment by nicholasjackson** [comment]
I have managed to get this working by using the following solution with podman-desktop 1.19.2, podman-machine 5.5.1 (wsl2) ubuntu 24.04 (wsl2).

1. In the *main wsl instance* where you want to share volumes with podman, add a system unit that mounts the root of the ubuntu instance to `/mnt/instances/<distro-name>`

```bash
cat <<EOF | sudo tee /etc/systemd/system/mnt-wsl-instances-${WSL_DISTRO_NAME/-/}.mount
[Unit]
Description=WSL Instances

[Mount]
What=/
Where=/mnt/wsl/instances/${WSL_DISTRO_NAME/-/}
Type=none
Options=defaults,bind,X-mount.mkdir

[Install]
WantedBy=multi-user.target
EOF

sudo systemctl daemon-reload

sudo systemctl enable mnt-wsl-instances-${WSL_DISTRO_NAME/-/}.mount --now
```

2. In the *podman machine* simlink a subfolder from the `/mnt/wsl/instances/<distro>` to the path as it would be on the *main wsl instance*.  For example, I simlink my home directory in the *main wsl instance* `/home/nicj` to the podman machine `/home/nicj`.

```bash
sudo ln -s /mnt/wsl/instances/ubuntu/home/nicj /home/nicj
```

3. You should now be able to do `podman run -v ${PWD}:/files --rm -i alpine:latest` and as long as ${PWD} is within your home folder then the mount will work as the paths are the same on both machines.

I use my home folder as it is where I store all my files but as long as you create a symlink on the podman-machine that is the same location as the files on the main wsl instance you can use any folders you like.

**Reply by GenosW** [reply]
Hi,
first up: thank you so much!

Secondly, there's one minor wrinkle you might run into IF you've setup a custom user in your other WSL distros like me - for whatever reason I had back then. I don't want to migrate this WSL distro setup to one with UID=1000 user for various reasons (mostly effort related...).
You'll end up with your WSL custom users having some UID > 1000 (e.g. 1001) while the podman machine user will have 1000. rootless podman will be unable mount any volumes within `/mnt/wsl/instances/<distro>` due to permissions and UID mismatch.

I've not found a good solution for this issue...other than using the rootful podman connection of course.

**Comment by carlosedp** [comment]
When I need to mount local paths to the containers, I end up unsetting the CONTAINER_HOST to use the local WSL podman instance instead of the shared one that's on the podman-machine (which is another WSL instance).
